File: merge_sort.py


def merge(ls1, ls2):
	# print("\nlist inputs to merge: {} {}".format(ls1, ls2))
	ordered_list = []
	
	i = 0 # counter for ls1
	j = 0 # counter for ls2
	while(i < len(ls1) and j < len(ls2)):
		# do stuff until one of them runs out
		if(ls1[i] < ls2[j]):
			ordered_list = ordered_list + [ls1[i]]
			i += 1
		else:
			ordered_list = ordered_list + [ls2[j]]
			j += 1
	
	if(i < len(ls1)):
		leftover = ls1[i:]
	else:
		leftover = ls2[j:]

	return ordered_list + leftover


def merge_sort(ls):
	if len(ls) == 1:
		return ls

	mid = len(ls)/2
	array_one = ls[:mid]
	array_two = ls[mid:]

	array_one = merge_sort(array_one)
	array_two = merge_sort(array_two)

	return merge(array_one, array_two)


if __name__ == '__main__':
    import numpy as np
    x = np.arange(21)
    np.random.shuffle(x)
    print x.tolist()
    quick_sort(x.tolist())

File: scala/mergeSort.scala


def merge(ls1: List[Int], ls2: List[Int]): List[Int] = {
  println(s"\nlist inputs to split: $ls1 $ls2")
  var orderedList = List[Int]()
  var leftOver = List[Int]()

  var i = 0 // counter for ls1
  var j = 0 // counter for ls2
  while(i < ls1.length && j < ls2.length) {
    // do stuff until one of them (i or j) runs out
    if(ls1(i) < ls2(j)) {
      orderedList = orderedList ::: List(ls1(i))
      i = i + 1
    } else {
      orderedList = orderedList ::: List(ls2(j))
      j = j + 1
    }
  }
  
  if(i < ls1.length) leftOver = ls1.slice(i, ls1.length)
  else leftOver = ls2.slice(j, ls2.length)

  println(s"merged list: ${orderedList ::: leftOver}")
  return orderedList ::: leftOver
}

def mergeSort(ls: List[Int]): List[Int] = {
  if(ls.length == 1) return ls

  val mid = ls.length/2
  var arrayOne = ls.slice(0, mid)
  var arrayTwo = ls.slice(mid, ls.length)
  println(s"\nsplit $ls")

  arrayOne = mergeSort(arrayOne)
  arrayTwo = mergeSort(arrayTwo)

  val orderedSublist = merge(arrayOne, arrayTwo)
  return orderedSublist
}
mergeSort(List(2,8,5,3,9,4,70,8))

